File: src/main/java/com/pubnub/api/managers/StateManager.java
package com.pubnub.api.managers;

import com.pubnub.api.PubNub;
import com.pubnub.api.builder.dto.PresenceOperation;
import com.pubnub.api.builder.dto.StateOperation;
import com.pubnub.api.builder.dto.SubscribeOperation;
import com.pubnub.api.builder.dto.UnsubscribeOperation;
import com.pubnub.api.models.SubscriptionItem;
import lombok.AllArgsConstructor;
import lombok.Data;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class StateManager {
    @Data
    @AllArgsConstructor
    private static class TemporaryUnavailableItem {
        private String item;
        private Date timestamp;
    }

    static final int MILLIS_IN_SECOND = 1000;

    /**
     * Contains a list of subscribed channels
     */
    private Map<String, SubscriptionItem> channels;
    /**
     * Contains a list of subscribed presence channels.
     */
    private Map<String, SubscriptionItem> presenceChannels;

    /**
     * Contains a list of subscribed channel groups.
     */
    private Map<String, SubscriptionItem> groups;

    /**
     * Contains a list of subscribed presence channel groups.
     */
    private Map<String, SubscriptionItem> presenceGroups;

    private Map<String, SubscriptionItem> heartbeatChannels;
    private Map<String, SubscriptionItem> heartbeatGroups;

    private List<TemporaryUnavailableItem> temporaryUnavailableChannels = new ArrayList<>();
    private List<TemporaryUnavailableItem> temporaryUnavailableChannelGroups = new ArrayList<>();

    private final PubNub pubnub;

    public StateManager(final PubNub pubnub) {
        this.pubnub = pubnub;
        this.channels = new HashMap<>();
        this.presenceChannels = new HashMap<>();

        this.groups = new HashMap<>();
        this.presenceGroups = new HashMap<>();

        this.heartbeatChannels = new HashMap<>();
        this.heartbeatGroups = new HashMap<>();
    }

    public synchronized void adaptSubscribeBuilder(SubscribeOperation subscribeOperation) {
        for (String channel : subscribeOperation.getChannels()) {
            if (channel == null || channel.length() == 0) {
                continue;
            }

            SubscriptionItem subscriptionItem = new SubscriptionItem().setName(channel);
            channels.put(channel, subscriptionItem);

            if (subscribeOperation.isPresenceEnabled()) {
                SubscriptionItem presenceSubscriptionItem = new SubscriptionItem().setName(channel);
                presenceChannels.put(channel, presenceSubscriptionItem);
            }

        }

        for (String channelGroup : subscribeOperation.getChannelGroups()) {
            if (channelGroup == null || channelGroup.length() == 0) {
                continue;
            }

            SubscriptionItem subscriptionItem = new SubscriptionItem().setName(channelGroup);
            groups.put(channelGroup, subscriptionItem);

            if (subscribeOperation.isPresenceEnabled()) {
                SubscriptionItem presenceSubscriptionItem = new SubscriptionItem().setName(channelGroup);
                presenceGroups.put(channelGroup, presenceSubscriptionItem);
            }

        }
    }

    public synchronized void adaptStateBuilder(StateOperation stateOperation) {
        for (String channel : stateOperation.getChannels()) {
            SubscriptionItem subscribedChannel = channels.get(channel);

            if (subscribedChannel != null) {
                subscribedChannel.setState(stateOperation.getState());
            }
        }

        for (String channelGroup : stateOperation.getChannelGroups()) {
            SubscriptionItem subscribedChannelGroup = groups.get(channelGroup);

            if (subscribedChannelGroup != null) {
                subscribedChannelGroup.setState(stateOperation.getState());
            }
        }
    }


    public synchronized void adaptUnsubscribeBuilder(UnsubscribeOperation unsubscribeOperation) {
        for (String channel : unsubscribeOperation.getChannels()) {
            this.channels.remove(channel);
            this.presenceChannels.remove(channel);
            removeTemporaryUnavailableChannel(channel);
        }

        for (String channelGroup : unsubscribeOperation.getChannelGroups()) {
            this.groups.remove(channelGroup);
            this.presenceGroups.remove(channelGroup);
        }
    }

    public synchronized void adaptPresenceBuilder(PresenceOperation presenceOperation) {
        for (String channel : presenceOperation.getChannels()) {
            if (channel == null || channel.length() == 0) {
                continue;
            }

            if (presenceOperation.isConnected()) {
                SubscriptionItem subscriptionItem = new SubscriptionItem().setName(channel);
                heartbeatChannels.put(channel, subscriptionItem);
            } else {
                heartbeatChannels.remove(channel);
            }

        }

        for (String channelGroup : presenceOperation.getChannelGroups()) {
            if (channelGroup == null || channelGroup.length() == 0) {
                continue;
            }

            if (presenceOperation.isConnected()) {
                SubscriptionItem subscriptionItem = new SubscriptionItem().setName(channelGroup);
                heartbeatGroups.put(channelGroup, subscriptionItem);
            } else {
                heartbeatGroups.remove(channelGroup);
            }

        }
    }

    public synchronized Map<String, Object> createStatePayload() {
        Map<String, Object> stateResponse = new HashMap<>();

        for (SubscriptionItem channel : channels.values()) {
            if (channel.getState() != null) {
                stateResponse.put(channel.getName(), channel.getState());
            }
        }

        for (SubscriptionItem channelGroup : groups.values()) {
            if (channelGroup.getState() != null) {
                stateResponse.put(channelGroup.getName(), channelGroup.getState());
            }
        }

        return stateResponse;
    }

    public synchronized List<String> prepareTargetChannelList(boolean includePresence) {
        return prepareMembershipList(channels, presenceChannels, includePresence);
    }

    public synchronized List<String> prepareTargetChannelGroupList(boolean includePresence) {
        return prepareMembershipList(groups, presenceGroups, includePresence);
    }

    public synchronized List<String> prepareTargetHeartbeatChannelList(boolean includePresence) {
        return prepareMembershipList(heartbeatChannels, presenceChannels, includePresence);
    }

    public synchronized List<String> prepareTargetHeartbeatChannelGroupList(boolean includePresence) {
        return prepareMembershipList(heartbeatGroups, presenceGroups, includePresence);
    }

    public boolean hasAnythingToSubscribe() {
        final List<String> combinedChannels = prepareTargetChannelList(true);
        final List<String> combinedChannelGroups = prepareTargetChannelGroupList(true);

        return !combinedChannels.isEmpty() || !combinedChannelGroups.isEmpty();
    }

    public synchronized void resetTemporaryUnavailableChannelsAndGroups() {
        temporaryUnavailableChannels.clear();
        temporaryUnavailableChannelGroups.clear();
    }

    public synchronized void removeTemporaryUnavailableChannel(String channel) {
        TemporaryUnavailableItem temporaryUnavailableChannelsToBeRemoved = null;
        for (final TemporaryUnavailableItem temporaryUnavailableChannel : temporaryUnavailableChannels) {
            if (temporaryUnavailableChannel.getItem().equals(channel)) {
                temporaryUnavailableChannelsToBeRemoved = temporaryUnavailableChannel;
            }
        }
        if (temporaryUnavailableChannelsToBeRemoved != null) {
            temporaryUnavailableChannels.remove(temporaryUnavailableChannelsToBeRemoved);
        }
    }

    public synchronized void removeTemporaryUnavailableChannelGroup(String channelGroup) {
        TemporaryUnavailableItem temporaryUnavailableChannelGroupsToBeRemoved = null;
        for (final TemporaryUnavailableItem temporaryUnavailableChannelGroup : temporaryUnavailableChannelGroups) {
            if (temporaryUnavailableChannelGroup.getItem().equals(channelGroup)) {
                temporaryUnavailableChannelGroupsToBeRemoved = temporaryUnavailableChannelGroup;
            }
        }
        if (temporaryUnavailableChannelGroupsToBeRemoved != null) {
            temporaryUnavailableChannelGroups.remove(temporaryUnavailableChannelGroupsToBeRemoved);
        }
    }

    public synchronized void addTemporaryUnavailableChannel(String channel) {
        temporaryUnavailableChannels.add(new TemporaryUnavailableItem(channel, new Date()));
    }

    public synchronized void addTemporaryUnavailableChannelGroup(String channelGroupName) {
        temporaryUnavailableChannelGroups.add(new TemporaryUnavailableItem(channelGroupName, new Date()));
    }

    public boolean subscribedToOnlyTemporaryUnavailable() {
        return effectiveChannels().isEmpty() && effectiveChannelGroups().isEmpty();
    }

    public List<String> effectiveChannels() {
        final List<String> effectiveChannelsList = prepareTargetChannelList(true);
        effectiveChannelsList.removeAll(channelsToPostponeSubscription(temporaryUnavailableChannels));
        return effectiveChannelsList;
    }

    public List<String> effectiveChannelGroups() {
        final List<String> effectiveChannelGroupsList = prepareTargetChannelGroupList(true);
        effectiveChannelGroupsList.removeAll(channelGroupsToPostponeSubscription(temporaryUnavailableChannelGroups));
        return effectiveChannelGroupsList;
    }

    private List<String> channelsToPostponeSubscription(final List<TemporaryUnavailableItem> temporaryUnavailableChannels) {
        final List<String> result = new ArrayList<>();

        for (TemporaryUnavailableItem temporaryUnavailableChannel : temporaryUnavailableChannels) {
            if (temporaryUnavailableChannel.getTimestamp()
                    .after(new Date(System.currentTimeMillis() - pubnub.getConfiguration().getConnectTimeout() * MILLIS_IN_SECOND))) {
                result.add(temporaryUnavailableChannel.getItem());
            }
        }
        return result;
    }

    private List<String> channelGroupsToPostponeSubscription(final List<TemporaryUnavailableItem> temporaryUnavailableChannelGroups) {
        final List<String> result = new ArrayList<>();

        for (TemporaryUnavailableItem temporaryUnavailableChannelGroup : temporaryUnavailableChannelGroups) {
            if (temporaryUnavailableChannelGroup.getTimestamp()
                    .after(new Date(System.currentTimeMillis() - pubnub.getConfiguration().getConnectTimeout() * MILLIS_IN_SECOND))) {
                result.add(temporaryUnavailableChannelGroup.getItem());
            }
        }
        return result;
    }

    public synchronized boolean isEmpty() {
        return (channels.isEmpty() && presenceChannels.isEmpty() && groups.isEmpty() && presenceGroups.isEmpty());
    }

    private synchronized List<String> prepareMembershipList(Map<String, SubscriptionItem> dataStorage, Map<String,
            SubscriptionItem> presenceStorage, boolean includePresence) {
        List<String> response = new ArrayList<>();

        for (SubscriptionItem channelGroupItem : dataStorage.values()) {
            response.add(channelGroupItem.getName());
        }

        if (includePresence) {
            for (SubscriptionItem presenceChannelGroupItem : presenceStorage.values()) {
                response.add(presenceChannelGroupItem.getName().concat("-pnpres"));
            }
        }


        return response;
    }

}

File: src/main/kotlin/com/pubnub/api/managers/StateManager.kt
package com.pubnub.api.managers

import com.pubnub.api.builder.ConnectedStatusAnnouncedOperation
import com.pubnub.api.builder.NoOpOperation
import com.pubnub.api.builder.PresenceOperation
import com.pubnub.api.builder.PubSubOperation
import com.pubnub.api.builder.StateOperation
import com.pubnub.api.builder.SubscribeOperation
import com.pubnub.api.builder.TimetokenRegionOperation
import com.pubnub.api.builder.UnsubscribeOperation
import com.pubnub.api.models.SubscriptionItem
import com.pubnub.api.subscribe.PRESENCE_CHANNEL_SUFFIX

class StateManager {

    /**
     * Contains a list of subscribed channels
     */
    private val channels: HashMap<String, SubscriptionItem> = hashMapOf()

    /**
     * Contains a list of subscribed presence channels.
     */
    private val presenceChannels: HashMap<String, SubscriptionItem> = hashMapOf()

    /**
     * Contains a list of subscribed channel groups.
     */
    private val groups: HashMap<String, SubscriptionItem> = hashMapOf()

    /**
     * Contains a list of subscribed presence channel groups.
     */
    private val presenceGroups: HashMap<String, SubscriptionItem> = hashMapOf()

    private val heartbeatChannels: HashMap<String, SubscriptionItem> = hashMapOf()
    private val heartbeatGroups: HashMap<String, SubscriptionItem> = hashMapOf()

    /**
     * Keep track of Region to support PSV2 specification.
     */
    private var region: String? = null
    private var timetoken = 0L

    /**
     * Store the latest timetoken to subscribe with, null by default to get the latest timetoken.
     */
    private var storedTimetoken: Long? = null

    private var shouldAnnounce: Boolean = false

    @Synchronized
    internal fun handleOperation(vararg pubSubOperations: PubSubOperation) {
        pubSubOperations.forEach { pubSubOperation ->
            when (pubSubOperation) {
                is SubscribeOperation -> {
                    adaptSubscribeBuilder(subscribeOperation = pubSubOperation)
                    shouldAnnounce = true
                }
                is UnsubscribeOperation -> {
                    adaptUnsubscribeBuilder(unsubscribeOperation = pubSubOperation)
                    shouldAnnounce = true
                }
                is PresenceOperation -> adaptPresenceBuilder(presenceOperation = pubSubOperation)
                is StateOperation -> adaptStateBuilder(stateOperation = pubSubOperation)
                is TimetokenRegionOperation -> {
                    if (storedTimetoken != null) {
                        timetoken = storedTimetoken!!
                        storedTimetoken = null
                    } else {
                        timetoken = pubSubOperation.timetoken
                    }

                    region = pubSubOperation.region
                }
                ConnectedStatusAnnouncedOperation -> {
                    shouldAnnounce = false
                }
                NoOpOperation -> {
                    // do nothing
                }
            }
        }
    }

    @Synchronized
    internal fun subscriptionStateData(includePresence: Boolean): SubscriptionStateData {
        return SubscriptionStateData(
            channelGroups = prepareChannelGroupList(includePresence),
            channels = prepareChannelList(includePresence),
            heartbeatChannelGroups = prepareHeartbeatChannelGroupList(includePresence),
            heartbeatChannels = prepareHeartbeatChannelList(includePresence),
            timetoken = timetoken,
            region = region,
            statePayload = createStatePayload(),
            shouldAnnounce = shouldAnnounce
        )
    }

    private fun adaptSubscribeBuilder(subscribeOperation: SubscribeOperation) {
        for (channel in subscribeOperation.channels) {
            if (channel.isEmpty()) {
                continue
            }
            val subscriptionItem = SubscriptionItem(channel)

            channels[channel] = subscriptionItem

            if (subscribeOperation.presenceEnabled) {
                val presenceSubscriptionItem = SubscriptionItem(channel)
                presenceChannels[channel] = presenceSubscriptionItem
            }
        }
        for (channelGroup in subscribeOperation.channelGroups) {
            if (channelGroup.isEmpty()) {
                continue
            }
            val subscriptionItem = SubscriptionItem(channelGroup)
            groups.put(channelGroup, subscriptionItem)
            if (subscribeOperation.presenceEnabled) {
                val presenceSubscriptionItem = SubscriptionItem(channelGroup)
                presenceGroups[channelGroup] = presenceSubscriptionItem
            }
        }
        timetoken = subscribeOperation.timetoken

        // if the timetoken is not at starting position, reset the timetoken to get a connected event
        // and store the old timetoken to be reused later during subscribe.
        if (timetoken != 0L) {
            storedTimetoken = timetoken
        }
        timetoken = 0L
    }

    private fun adaptStateBuilder(stateOperation: StateOperation) {
        for (channel in stateOperation.channels) {
            val subscribedChannel = channels[channel]
            subscribedChannel?.state = stateOperation.state
        }
        for (channelGroup in stateOperation.channelGroups) {
            val subscribedChannelGroup = groups[channelGroup]
            subscribedChannelGroup?.state = stateOperation.state
        }
    }

    private fun adaptUnsubscribeBuilder(unsubscribeOperation: UnsubscribeOperation) {
        for (channel in unsubscribeOperation.channels) {
            channels.remove(channel)
            presenceChannels.remove(channel)
        }
        for (channelGroup in unsubscribeOperation.channelGroups) {
            groups.remove(channelGroup)
            presenceGroups.remove(channelGroup)
        }
        // if we unsubscribed from all the channels, reset the timetoken back to zero and remove the region.
        if (isEmpty()) {
            region = null
            storedTimetoken = null
            timetoken = 0L
        } else {
            storedTimetoken = timetoken
            timetoken = 0L
        }
    }

    private fun adaptPresenceBuilder(presenceOperation: PresenceOperation) {
        for (channel in presenceOperation.channels) {
            if (channel.isEmpty()) {
                continue
            }
            if (presenceOperation.connected) {
                val subscriptionItem = SubscriptionItem(channel)
                heartbeatChannels[channel] = subscriptionItem
            } else {
                heartbeatChannels.remove(channel)
            }
        }
        for (channelGroup in presenceOperation.channelGroups) {
            if (channelGroup.isEmpty()) {
                continue
            }
            if (presenceOperation.connected) {
                val subscriptionItem = SubscriptionItem(channelGroup)
                heartbeatGroups[channelGroup] = subscriptionItem
            } else {
                heartbeatGroups.remove(channelGroup)
            }
        }
    }

    private fun createStatePayload(): Map<String, Any?> {
        val stateResponse: HashMap<String, Any?> = hashMapOf()
        for (channel in channels.values) {
            if (channel.state != null) {
                stateResponse[channel.name] = channel.state
            }
        }
        for (channelGroup in groups.values) {
            if (channelGroup.state != null) {
                stateResponse[channelGroup.name] = channelGroup.state
            }
        }
        return stateResponse
    }

    private fun prepareChannelList(includePresence: Boolean): List<String> {
        return prepareMembershipList(channels, presenceChannels, includePresence)
    }

    private fun prepareChannelGroupList(includePresence: Boolean): List<String> {
        return prepareMembershipList(groups, presenceGroups, includePresence)
    }

    private fun prepareHeartbeatChannelList(includePresence: Boolean): List<String> {
        return prepareMembershipList(heartbeatChannels, presenceChannels, includePresence)
    }

    private fun prepareHeartbeatChannelGroupList(includePresence: Boolean): List<String> {
        return prepareMembershipList(heartbeatGroups, presenceGroups, includePresence)
    }

    private fun isEmpty(): Boolean {
        return channels.isEmpty() && presenceChannels.isEmpty() && groups.isEmpty() && presenceGroups.isEmpty()
    }

    private fun prepareMembershipList(
        dataStorage: Map<String, SubscriptionItem>,
        presenceStorage: Map<String, SubscriptionItem>,
        includePresence: Boolean
    ): List<String> {
        val response: MutableList<String> = ArrayList()
        for (channelGroupItem in dataStorage.values) {
            response.add(channelGroupItem.name)
        }
        if (includePresence) {
            for (presenceChannelGroupItem in presenceStorage.values) {
                response.add(presenceChannelGroupItem.name + PRESENCE_CHANNEL_SUFFIX)
            }
        }
        return response
    }
}

internal data class SubscriptionStateData(
    val statePayload: Map<String, Any?>,
    val heartbeatChannelGroups: List<String>,
    val heartbeatChannels: List<String>,
    val channelGroups: List<String>,
    val channels: List<String>,
    val timetoken: Long,
    val region: String?,
    val shouldAnnounce: Boolean
)
